File: Code/src/ShapeFun/LME.py
#!/home/migmolper2/anaconda3/bin/python3

import numpy as np
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
from matplotlib import cm
from matplotlib.ticker import LinearLocator, FixedLocator, FormatStrFormatter

"""
  Shape functions based in :
  " Local maximum-entropy approximation schemes : a seamless 
  bridge between finite elements and meshfree methods "
  by M.Arroyo and M.Ortiz, 2006.

  Here we employ the same nomenclature as in the paper. With the single
  different of the "l" variable wich represents the distances between the
  evaluation point and the neighborhood nodes.

  List of functions :
  - LME_lambda_NR
  - LME_fa
  - LME_p
  - LME_r
  - LME_J
  - LME_dp
"""

def LME_fa(l_a, lambda_a, Beta_a):
    """
    """
    return - Beta_a * np.dot(l_a, l_a) + np.dot(l_a,lambda_a)

def LME_p(l, lambda_a, Beta_a):
    """
    Get the value of the shape function "pa" (1 x neighborhood) in the
    neighborhood nodes.
    
    Input parameters :
    -> l : Matrix with the distances to the
    neighborhood nodes (neiborghood x dim).
    -> lambda : Initial value of the lagrange multipliers (1 x dim).
    -> Beta : Tunning parameter (scalar).
    """
    
    # Definition of some parameters
    N_a = np.shape(l)[0]
    N_dim = 2
    Z = 0.0
    Z_m1 = 0.0

    # Vector with the values of the shape-function in the nodes
    p = np.zeros([N_a])

    # Get Z and the numerator
    for a in range(0,N_a):
        la = l[a,:]
        p[a] = np.exp(LME_fa(la,lambda_a,Beta_a))
        Z += p[a]

    # Get the inverse of Z
    Z_m1 = 1/Z

    # Divide by Z and get the final value
    for a in range(0,N_a):
        p[a] *= Z_m1
      
    # Return the value of the shape function
    return p


def LME_r(l,p):
    """
    Get the gradient "r" (dim x 1) of the function log(Z) = 0.
    Input parameters :
    -> l : Matrix with the distances to the 
    neighborhood nodes (neighborhood x dim).
    -> p : Shape function value in the
    neighborhood nodes (1 x neighborhood).
    """
    
    # Definition of some parameters
    N_a = np.shape(l)[0]
    N_dim = 2

    # Value of the gradient
    r = np.zeros([N_dim])
    
    # Fill ''r''
    for i in range(0,N_dim):
        for a in range(0,N_a):
            r[i] += p[a]*l[a][i]
            
    # Return the value of the gradient
    return r

def LME_J(l, p, r):
    """
    Get the Hessian "J" (dim x dim) of the function log(Z) = 0.
    Input parameters :
    -> l : Matrix with the distances to the
    neighborhood nodes (neighborhood x dim).
    -> p : Shape function value in the
    neighborhood nodes (neighborhood x 1).
    -> r : Gradient of log(Z) (dim x 1).
    """  
    # Definition of some parameters
    N_a = np.shape(l)[0]
    N_dim = 2
  
    # Allocate Hessian */
    J = np.zeros([N_dim,N_dim])

    # Fill the Hessian
    # Get the first component of the Hessian
    for a in range(0,N_a):
        J = J + p[a]*np.tensordot(l[a,:],l[a,:],axes = 0)
    # Get the second value of the Hessian
    J = J - np.tensordot(r,r,axes = 0)
    
    # Return the value of the Hessian
    return J


def LME_dp(l, p):
    """
    Value of the shape function gradient "dp" (dim x neighborhood) in 
    the neighborhood nodes.
    Input parameters :
    -> l : Matrix with the distances to the
    neighborhood nodes (neighborhood x dim).
    -> p : Shape function value in the
    neighborhood nodes (neighborhood x 1).
    """
    
    # Definition of some parameters */
    N_a = np.shape(l)[0]
    N_dim = 2
    dp = np.zeros([N_dim,N_a])
  
    # Get the Gradient and the Hessian of log(Z) 
    r = LME_r(l,p)
    J = LME_J(l,p,r)
  
    # Inverse of the Hessian */
    Jm1 = np.linalg.inv(J)
  
    # Fill the gradient for each node */
    for a in range(0,N_a):
        la = l[a,:] 
        Jm1_la = np.dot(Jm1, la)    
        for i in range(0,N_dim):
            dp[i][a] = - p[a]*Jm1_la[i]
  
    # Return the value of the shape function gradient
    return dp


def LME_lambda_NR(l, lambda_GP, Beta):
    """
    Get the lagrange multipliers "lambda" (1 x dim) for the LME 
    shape function. The numerical method for that is the Newton-Rapson.
    
    Input parameters :
    -> l : Matrix with the distances to the
    neighborhood nodes (neighborhood x dim).
    -> lambda : Initial value of the
    lagrange multipliers (1 x dim).
    -> Beta : Tunning parameter (scalar).
    -> h : Grid spacing (scalar).
    -> TOL_zero : Tolerance for Newton-Rapson.
    """
    
    # Definition of some parameters */
    MaxIter = 100
    Ndim = 2
    NumIter = 0 # Iterator counter
    norm_r = 10.0 # Value of the norm
    TOL_NR = 10e-6 # tolerance
    
    # Start with the Newton-Rapson method
    while norm_r > TOL_NR :
	
        # Get vector with the shape functions evaluated in the nodes
        p = LME_p(l,lambda_GP,Beta)

        # Get the gradient of log(Z) and its norm
        r = LME_r(l,p)
        norm_r = np.linalg.norm(r)

        # Get the Hessian of log(Z) and update it with +||r||*I 
        # according with Dennis M.Kochmann et al. 2019 (CMAME)
        J = LME_J(l,p,r)
        for i in range(0,Ndim):
            J[i][i] += norm_r

        # Check the conditioning number of the Hessian
        if np.linalg.cond(J) > 10 :
            print("Error in LME_lambda_NR. ","Iter :",NumIter,
    	        " The Hessian is near to singular matrix")
            exit()
    
        # Get the increment of lambda
        D_lambda = np.linalg.solve(J,r)

        # Update the value of lambda
        for i  in range(0,Ndim):
            lambda_GP[i] -= D_lambda[i]
    
        # Update the number of iterations
        NumIter += 1
        if NumIter >= MaxIter :
            print("Warning in LME_lambda.","No convergence in",
                  NumIter,MaxIter,"iterations")
            print("Error",norm_r)
            exit()
    
  
    # Once the stopping criteria is reached, 
    # return the lagrange multipliers value
    return lambda_GP

N_gp = 35
N_n = 5
L = 2.
DX = (2*L-0.2)/N_n
Gamma = 17.3 # 17.3
Gamma = ([17.3,10.0,7.0,5.0])
Node = 12 # 12
format_fig='png'
# Show
Show_grafs = False

xp = yp = np.linspace(-L+0.1, L-0.1, N_gp)
zp = np.zeros([1,N_gp])
Xp, Yp = np.meshgrid(xp, yp, indexing='ij', sparse=False)
Zp = np.zeros([N_gp,N_gp])
N_I = np.zeros_like(Zp)
dNdx_I = np.zeros_like(Zp)
dNdy_I = np.zeros_like(Zp)

xI = yI = np.linspace(-L, L, N_n)
zI = np.zeros([1,N_n])
XI, YI = np.meshgrid(xI, yI, indexing='ij', sparse=False)
ZI = np.zeros([N_n,N_n])

l = np.zeros([N_n*N_n,2])

lambda_GP = np.zeros([N_gp*N_gp,2])
Xp_Yp = np.zeros([N_gp*N_gp,2])

for i in range(0,N_gp):
    for j in range(0,N_gp):
        Xp_Yp[i*N_gp + j][0] = Xp[i][j]
        Xp_Yp[i*N_gp + j][1] = Yp[i][j]

for Gamma_i in Gamma:

    # Get the value of gamma
    Beta = Gamma_i/(L*L)
    
    # Distance to the GPa
    for a in range(0,N_gp*N_gp):
        for i in range(0,N_n):
            for j in range(0,N_n):
                l[i*N_n + j][0] = Xp_Yp[a][0] - XI[i][j]
                l[i*N_n + j][1] = Xp_Yp[a][1] - YI[i][j]

        lambda_GP[a,:] = LME_lambda_NR(l, np.zeros([2]), Beta);

        # GPs
        # ax.scatter(XI, YI, ZI, c='r', marker='o')

    for a in range(0,N_gp*N_gp):
        for i in range(0,N_n):
            for j in range(0,N_n):
                l[i*N_n + j][0] = Xp_Yp[a][0] - XI[i][j]
                l[i*N_n + j][1] = Xp_Yp[a][1] - YI[i][j]
            
        N_I_GP = LME_p(l, lambda_GP[a,:], Beta)
        dN_I_GP = LME_dp(l, N_I_GP)
        i = int(a/(N_gp))
        j = a % N_gp
        N_I[i][j] = N_I_GP[Node]
        dNdx_I[i][j] = dN_I_GP[0][Node]
        dNdy_I[i][j] = dN_I_GP[1][Node]


    fig = plt.figure()
    ax = fig.gca(projection='3d')
    ax.grid(False)
    ax.xaxis.pane.set_edgecolor('black')
    ax.yaxis.pane.set_edgecolor('black')
    ax.zaxis.pane.set_edgecolor('black')
    ax.xaxis.pane.fill = False
    ax.yaxis.pane.fill = False
    ax.zaxis.pane.fill = False
    ax.axes.xaxis.set_ticklabels([])
    ax.axes.yaxis.set_ticklabels([])
    for line in ax.xaxis.get_ticklines():
        line.set_visible(False)
    for line in ax.yaxis.get_ticklines():
        line.set_visible(False)
    for line in ax.zaxis.get_ticklines():
        line.set_visible(False)
    surf = ax.plot_surface(Xp, Yp, N_I, rstride=1, cstride=1, cmap=cm.jet,
                           linewidth=0, antialiased=False)
    ax.set_zlim3d(0.00, 1.00)
    ax.set_zticks([0, 0.5, 1])
    plt.tight_layout()
    plt.savefig('LME_%1.1f_Shape_Fun.%s'%(Gamma_i,format_fig))  
    if Show_grafs:
        plt.show()
    fig.clear()


    fig = plt.figure()
    ax = fig.gca(projection='3d')
    ax.grid(False)
    ax.xaxis.pane.set_edgecolor('black')
    ax.yaxis.pane.set_edgecolor('black')
    ax.zaxis.pane.set_edgecolor('black')
    ax.xaxis.pane.fill = False
    ax.yaxis.pane.fill = False
    ax.zaxis.pane.fill = False
    ax.axes.xaxis.set_ticklabels([])
    ax.axes.yaxis.set_ticklabels([])
    ax.axes.zaxis.set_ticklabels([])
    for line in ax.xaxis.get_ticklines():
        line.set_visible(False)
    for line in ax.yaxis.get_ticklines():
        line.set_visible(False)
    for line in ax.zaxis.get_ticklines():
        line.set_visible(False)    
    surf = ax.plot_surface(Xp, Yp, dNdx_I, rstride=1, cstride=1, cmap=cm.jet,
                           linewidth=0, antialiased=False)
    ax.set_zlim3d(-1.00, 1.00)
    plt.tight_layout()
    plt.savefig('LME_%1.1f_Shape_Fun_dx.%s'%(Gamma_i,format_fig))  
    if Show_grafs:
        plt.show()
    fig.clear()
    
    fig = plt.figure()
    ax = fig.gca(projection='3d')
    ax.grid(False)
    ax.xaxis.pane.set_edgecolor('black')
    ax.yaxis.pane.set_edgecolor('black')
    ax.zaxis.pane.set_edgecolor('black')
    ax.xaxis.pane.fill = False
    ax.yaxis.pane.fill = False
    ax.zaxis.pane.fill = False
    ax.axes.xaxis.set_ticklabels([])
    ax.axes.yaxis.set_ticklabels([])
    ax.axes.zaxis.set_ticklabels([])
    for line in ax.xaxis.get_ticklines():
        line.set_visible(False)
    for line in ax.yaxis.get_ticklines():
        line.set_visible(False)
    for line in ax.zaxis.get_ticklines():
        line.set_visible(False)
    surf = ax.plot_surface(Xp, Yp, dNdy_I, rstride=1, cstride=1, cmap=cm.jet,
                        linewidth=0, antialiased=False)
    ax.set_zlim3d(-1.00, 1.00)
    plt.tight_layout()
    plt.savefig('LME_%1.1f_Shape_Fun_dy.%s'%(Gamma_i,format_fig))  
    if Show_grafs:
        plt.show()
    fig.clear()

File: Code/src/ShapeFun/LME.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <string.h>
#include "grams.h"

/***********************************************/
/************* Local Maximum-Entropy **************/
/**************************************************/

/*!
  Shape functions based in :
  "" Local maximum-entropy approximation schemes : a seamless 
  bridge between finite elements and meshfree methods ""
  by M.Arroyo and M.Ortiz, 2006.

  Here we employ the same nomenclature as in the paper. With the single
  different of the "l" variable wich represents the distances between the
  evaluation point and the neighborhood nodes.

  List of functions :
  - LME_Init_lambda
  - LME_lambda_NR
  - LME_fa
  - LME_p
  - LME_r
  - LME_J
  - LME_dp
  - LME_Tributary_Nodes
*/


/****************************************************************************/

void LME_Initialize_Beta(Matrix Beta, double DeltaX, int Np)
/*!
  Function to get a initial value of Beta
*/
{
  int Ndim = NumberDimensions;
  
  for(int p = 0 ; p<Np ; p++){
    for(int i = 0 ; i<Ndim ; i++){
      Beta.nM[p][i] = gamma_LME/(DeltaX*DeltaX);
    }
  }
  
}

/****************************************************************************/

void LME_Initialize(GaussPoint MPM_Mesh, Mesh FEM_Mesh)
{

  int Ndim = NumberDimensions;
  int Np = MPM_Mesh.NumGP;
  int Nelem = FEM_Mesh.NumElemMesh;

  /* Variables for the GP coordinates */  
  Matrix X_p = MatAssign(Ndim,1,NAN,NULL,NULL);


  Element Elem_p;
  
  ChainPtr Nodes_p;
  int I_p;

  /* Variables for the poligon */
  int NumVertex;
  int * Poligon_Connectivity;
  Matrix Poligon_Coordinates;
  ChainPtr ListNodes_I;

  /* Auxiliar variables for LME */
  Matrix lambda_p = /* Lagrange multipliers */
    MatAssign(Ndim,1,NAN,NULL,NULL);

  /* Distance from GP to the nodes */
  Matrix Delta_Xip; 
  Matrix Distance_Xip;
  
  Matrix Beta_p = /* Tunning parameter */
      MatAssign(Ndim,1,NAN,NULL,NULL);
  int NumNodes_GP; /* Number of neibourghs */

  /* */
  int * List;

  /* Initialize Beta */
  LME_Initialize_Beta(MPM_Mesh.Beta, FEM_Mesh.DeltaX, Np);

  /* Loop over the particle mesh */
  for(int p = 0 ; p<Np ; p++){

    /* Get some properties for each particle */ 
    X_p.nV = MPM_Mesh.Phi.x_GC.nM[p];
    Beta_p.nV = MPM_Mesh.Beta.nM[p];
    /* Initialize lambda */
    lambda_p.nV = MPM_Mesh.lambda.nM[p];

    /* Loop over the element mesh */
    for(int i = 0 ; i<Nelem ; i++){

      /* Get the element properties */
      Elem_p = get_Element(p,FEM_Mesh.Connectivity[i],FEM_Mesh.NumNodesElem[i]);
      
      /* 6º Check out if the GP is in the Element */
      if(InOut_Element(X_p, Elem_p, FEM_Mesh.Coordinates)){

	/* Define initial list of nodes */
	Delta_Xip = get_set_Coordinates(Elem_p.Connectivity,
					X_p, FEM_Mesh.Coordinates);

	/*  */
	I_p = get_node_near_to(Delta_Xip,);

	/* Get the initial connectivity of the particle */
	MPM_Mesh.ListNodes[p] = LME_Tributary_Nodes(X_p,Beta_p,I_p,FEM_Mesh);

	/* Measure the size of the connectivity */
	MPM_Mesh.NumberNodes[p] = get_Lenght_Set(MPM_Mesh.ListNodes[p]);
       	
	/* Calculate distance from particle to each node in the neibourhood */
	Delta_Xip = get_set_Coordinates(MPM_Mesh.ListNodes[p],
					X_p, FEM_Mesh.Coordinates);
	Distance_Xip = get_nurbs_distance(Delta_Xip);

	/* Update the value of beta */
	Beta_p = LME_Beta(Beta_p, Delta_Xip, gamma_LME);
	
	/* Asign to each particle the closest node in the mesh
	 and to this node asign the particle */
	MPM_Mesh.I0[p] = get_node_near_to(Delta_Xip,);
	push_to_Set(&FEM_Mesh.I_particles[MPM_Mesh.I0[p]],p);

	/* Active those nodes that interact with the GP */
	ListNodes_I = MPM_Mesh.ListNodes[p];
	while(ListNodes_I != NULL){
	  FEM_Mesh.NumParticles[ListNodes_I->I] += 1;
	  ListNodes_I = ListNodes_I->next; 
	}
		
	/* Update lagrange multipliers with Newton-Rapson */
	lambda_p = LME_lambda_NR(Delta_Xip, lambda_p, Beta_p);

	/* Free memory */
	FreeMat(Delta_Xip);

	break;
	
      }
      
      /* 11º Free memory */
      free(Poligon_Connectivity);
      FreeMat(Poligon_Coordinates);
      
    }

  }

}

/****************************************************************************/

Matrix LME_Beta(Matrix Beta, Matrix l, double Gamma)
/*!
  Function to update the value of beta
*/
{

  int Ndim = NumberDimensions;
  int NumNodes_GP = l.N_rows;
  double h = 0;
  Matrix l_GP_I = MatAssign(Ndim,1,NAN,NULL,NULL);
  
  /* Get the mean distande */
  for(int i = 0 ; i<NumNodes_GP ; i++){
    l_GP_I.nV = l.nM[i];
    h += Norm_Mat(l_GP_I,2);
  }
  h = h/NumNodes_GP;

  /* Fill Beta */
  for(int j = 0 ; j<Ndim ; j++){
    Beta.nV[j] = Gamma/(h*h);
  }

  return Beta;
}

/****************************************************************************/

Matrix LME_lambda_NR(Matrix l, Matrix lambda, Matrix Beta)
/*!
  Get the lagrange multipliers "lambda" (1 x dim) for the LME 
  shape function. The numerical method for that is the Newton-Rapson.

  Input parameters :
  -> l : Matrix with the distances to the
  neighborhood nodes (neighborhood x dim).
  -> lambda : Initial value of the
  lagrange multipliers (1 x dim).
  -> Beta : Tunning parameter (scalar).
  -> h : Grid spacing (scalar).
  -> TOL_zero : Tolerance for Newton-Rapson.
*/
{
  /* Definition of some parameters */
  int MaxIter = 100;
  int Ndim = NumberDimensions;
  int NumIter = 0; /* Iterator counter */
  double norm_r = 10; /* Value of the norm */
  Matrix p; /* Shape function vector */
  Matrix r; /* Gradient of log(Z) */
  Matrix J; /* Hessian of log(Z) */
  Matrix D_lambda; /* Increment of lambda */
 
  /* Start with the Newton-Rapson method */
  while(norm_r > TOL_NR){
	
    /* Get vector with the shape functions evaluated in the nodes */
    p = LME_p(l,lambda,Beta);

    /* Get the gradient of log(Z) and its norm */
    r = LME_r(l,p), norm_r = Norm_Mat(r,2);

    /* Get the Hessian of log(Z) and update it with +||r||*I 
       according with Dennis M.Kochmann et al. 2019 (CMAME) */    
    J = LME_J(l,p,r);
    for(int i = 0 ; i<Ndim ; i++){
      J.nM[i][i] += norm_r;
    }

    /* /\* Check the conditioning number of the Hessian *\/ */
    /* if (fabs(Cond_Mat(J,TOL_NR)) > 10){ */
    /*   printf(" %s (%s %i) : %s \n", */
    /* 	     "Error in LME_lambda_NR","Iter",NumIter, */
    /* 	     "The Hessian is near to singular matrix"); */
    /*   exit(0); */
    /* } */
    
    /* Get the increment of lambda */
    D_lambda = Solve_Linear_Sistem(J,r,MatAllocZ(Ndim,1));

    /* Update the value of lambda */
    for(int i = 0 ; i<Ndim ; i++){
      lambda.nV[i] -= D_lambda.nV[i];
    }

    /* Free memory */
    FreeMat(p), FreeMat(r), FreeMat(J), FreeMat(D_lambda);
    
    /* Update the number of iterations */
    NumIter ++;
    if(NumIter >= MaxIter){
      printf("%s : %s %i/%i %s \n",
      	     "Warning in LME_lambda",
      	     "No convergence in",NumIter,MaxIter,"iterations");
      printf("%s : %f \n",
	     "Error",norm_r);
      exit(0);
    }
  
  }
  
  /* Once the stopping criteria is reached, 
     return the lagrange multipliers value */
  return lambda;
}

/****************************************************************************/

double LME_fa(Matrix la, Matrix lambda, Matrix Beta)
/*!
  Output :
  -> fa : the function fa that appear in [1] (scalar).
  Input parameters :
  -> la : Matrix with the distance to the neighborhood node ''a'' (1 x dim).
  -> lambda : Initial value of the lagrange multipliers (dim x 1).
  -> Gamma : Tunning parameter (scalar).
*/
{  
  int Ndim = NumberDimensions;
  double fa = 0;

  for(int i = 0 ; i<Ndim ; i++){
    fa += - Beta.nV[i]*la.nV[i]*la.nV[i] + la.nV[i]*lambda.nV[i];
  }
    
  /* Return the value of fa */
  return fa;
}

/****************************************************************************/

Matrix LME_p(Matrix l, Matrix lambda, Matrix Beta)
/*!
  Get the value of the shape function "pa" (1 x neighborhood) in the
  neighborhood nodes.

  Input parameters :
  -> l : Matrix with the distances to the
  neighborhood nodes (neiborghood x dim).
  -> lambda : Initial value of the lagrange multipliers (1 x dim).
  -> Beta : Tunning parameter (scalar).
*/
{
  
  /* Definition of some parameters */
  int N_a = l.N_rows;
  int Ndim = NumberDimensions;
  double Z = 0, Z_m1 = 0;
  Matrix p = /* Vector with the values of the shape-function in the nodes */
    MatAlloc(1,N_a); 
  Matrix la = /* Distance to the neighbour (x-x_a) */
    MatAssign(1,Ndim,NAN,NULL,NULL);

  /* Get Z and the numerator */
  for(int a = 0 ; a<N_a ; a++){
    la.nV = l.nM[a];
    p.nV[a] = exp(LME_fa(la,lambda,Beta));
    Z += p.nV[a];
  }

  /* Get the inverse of Z */
  Z_m1 = (double)1/Z;

  /* Divide by Z and get the final value */
  for(int a = 0 ; a<N_a ; a++){
    p.nV[a] *= Z_m1;
  }
  
  /* Return the value of the shape function */  
  return p;
}

/****************************************************************************/

Matrix LME_r(Matrix l, Matrix p)
/*!
  Get the gradient "r" (dim x 1) of the function log(Z) = 0.
  Input parameters :
  -> l : Matrix with the distances to the 
  neighborhood nodes (neighborhood x dim).
  -> p : Shape function value in the
  neighborhood nodes (1 x neighborhood).
*/
{  
  /* Definition of some parameters */
  int N_a = l.N_rows;
  int Ndim = NumberDimensions;
  Matrix r /* Value of the gradient */
    = MatAllocZ(Ndim,1);

  /* Fill ''r'' */
  for(int i = 0 ; i<Ndim ; i++){
    for(int a = 0 ; a<N_a ; a++){
      r.nV[i] += p.nV[a]*l.nM[a][i];
    }
  }

  /* Return the value of the gradient */
  return r;
}

/****************************************************************************/

Matrix LME_J(Matrix l, Matrix p, Matrix r)
/*!
  Get the Hessian "J" (dim x dim) of the function log(Z) = 0.
  Input parameters :
  -> l : Matrix with the distances to the
  neighborhood nodes (neighborhood x dim).
  -> p : Shape function value in the
  neighborhood nodes (neighborhood x 1).
  -> r : Gradient of log(Z) (dim x 1).
*/
{  
  /* Definition of some parameters */
  int N_a = l.N_rows;
  int Ndim = NumberDimensions;
  Matrix J; /* Hessian definition */
  
  /* Allocate Hessian */
  J = MatAllocZ(Ndim,Ndim);

  /* Fill the Hessian */
  for(int i = 0 ; i<Ndim ; i++){
    for(int j = 0 ; j<Ndim ; j++){
      for(int a = 0 ; a<N_a ; a++){
	/* Get the first component of the Hessian looping
	   over the neighborhood nodes. */
	J.nM[i][j] += p.nV[a]*l.nM[a][i]*l.nM[a][j];
      }
      /* Get the second value of the Hessian */
      J.nM[i][j] -= r.nV[i]*r.nV[j];
    }
  }

  /* Return the value of the Hessian */
  return J;
}

/****************************************************************************/

Matrix LME_dp(Matrix l, Matrix p)
/*!
  Value of the shape function gradient "dp" (dim x neighborhood) in 
  the neighborhood nodes.
  Input parameters :
  -> l : Matrix with the distances to the
  neighborhood nodes (neighborhood x dim).
  -> p : Shape function value in the
  neighborhood nodes (neighborhood x 1).
*/
{  
  /* Definition of some parameters */
  int N_a = l.N_rows;
  int Ndim = NumberDimensions;
  Matrix dp = MatAllocZ(N_a,Ndim);
  Matrix r; /* Gradient of log(Z) */
  Matrix J; /* Hessian of log(Z) */
  Matrix Jm1; /* Inverse of J */
  Matrix Jm1_la; /* Auxiliar vector */
  Matrix la = /* Distance to the neighbour (x-x_a) */
    MatAssign(Ndim,1,NAN,NULL,NULL); 
  
  /* Get the Gradient and the Hessian of log(Z) */
  r = LME_r(l,p);
  J = LME_J(l,p,r);
  
  /* Inverse of the Hessian */
  Jm1 = Get_Inverse(J);
  
  /* Free memory */
  FreeMat(r), FreeMat(J);

  /* Fill the gradient for each node */
  for(int a = 0 ; a<N_a ; a++){
    la.nV = l.nM[a]; 
    Jm1_la = Scalar_prod(Jm1,la);    
    for(int i = 0 ; i<Ndim ; i++){
      dp.nM[a][i] = - p.nV[a]*Jm1_la.nV[i];
    }
    FreeMat(Jm1_la);
  }

  /* Free memory */
  FreeMat(Jm1);
  
  /* Return the value of the shape function gradient */  
  return dp;  
}

/****************************************************************************/

ChainPtr LME_Tributary_Nodes(Matrix X_GP, Matrix Beta, int I0, Mesh FEM_Mesh){

  /* Define output */
  ChainPtr Triburary_Nodes = NULL;
  /* Number of dimensionws of the problem */
  int Ndim = NumberDimensions;

  Matrix Distance; /* Distance between node and GP */
  Matrix X_I = MatAssign(Ndim,1,NAN,NULL,NULL);
  
  ChainPtr Set_Nodes0 = NULL;
  int * Array_Nodes0;
  int NumNodes0;
  int Node0;

  /* Get the search radius */
  double Ra = sqrt(-log(TOL_lambda)/Beta.nV[0]);

  /* Get nodes close to the GP */
  Set_Nodes0 = get_NodesClose_toNode(I0, FEM_Mesh);
  NumNodes0 = get_Lenght_Set(Set_Nodes0);
  Array_Nodes0 = Set_to_Pointer(Set_Nodes0,NumNodes0);
  free_Set(Set_Nodes0);
     
  /* Loop over the chain with the tributary nodes */
  for(int i = 0 ; i<NumNodes0 ; i++){

    Node0 = Array_Nodes0[i];

    /* Assign to a pointer the coordinates of the nodes */
    X_I.nV = FEM_Mesh.Coordinates.nM[Node0];

    /* Get a vector from the GP to the node */
    Distance = Sub_Mat(X_GP,X_I);

    /* If the node is near the GP push in the chain */
    if(Norm_Mat(Distance,2) <= Ra){
      push_to_Set(&Triburary_Nodes,Node0);
    }

    /* Free memory of the distrance vector */
    FreeMat(Distance);

  }
  
  /* Free memory */
  free(Array_Nodes0);

  return Triburary_Nodes;
}

